File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 class Solution
 {
     public List<Integer> postorderTraversal(TreeNode root)
     {
         List<Integer> values = new ArrayList<>();
 
         if (root == null) return values;
 
         List<TreeNode> stackRoot = new ArrayList<>();
         stackRoot.add(root);
         List<TreeNode> stackNode = new ArrayList<>();
 
         while (!stackRoot.isEmpty())
         {
             var node = stackRoot.getLast();
 
             stackNode.add(node);
             stackRoot.removeLast();
 
             if (node.left != null)
                 stackRoot.add(node.left);
 
             if (node.right != null)
                 stackRoot.add(node.right);
         }
 
         while (!stackNode.isEmpty())
         {
             var node = stackNode.getLast();
 
             values.add(node.val);
             stackNode.removeLast();
         }
 
         return values;
     }
 }

File: leetcode/100-999/145. Binary Tree Postorder Traversal/solution_2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
 
var postorderTraversal = function(root) 
{
    const values = []

    if (!root) return values

    const stackRoot = [root]
    const stackNode = []

    while (stackRoot.length > 0)
    {
        const node = stackRoot.pop()

        stackNode.push(node)

        if (node.left)
            stackRoot.push(node.left)

        if (node.right)
            stackRoot.push(node.right)
    }

    while (stackNode.length > 0)
    {
        const node = stackNode.pop()
        values.push(node.val)
    }       

    return values
};